SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/SubViewTest.cpp

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialView/OpList/EntityComponentOpList.h"
+#include "SpatialView/ViewCoordinator.h"
+#include "SpatialViewUtils.h"
+#include "Tests/TestDefinitions.h"
+#include "Utils/ComponentFactory.h"
+
+#define SUBVIEW_TEST(TestName) GDK_TEST(Core, SubView, TestName)
+
+namespace SpatialGDK
+{
+const Worker_EntityId ENTITY_ID = 1;
+const Worker_EntityId TAGGED_ENTITY_ID = 2;
+const Worker_EntityId OTHER_TAGGED_ENTITY_ID = 3;
+const Worker_ComponentId TAG_COMPONENT_ID = 1;
+const Worker_ComponentId VALUE_COMPONENT_ID = 2;
+const double CORRECT_VALUE = 1;
+const double INCORRECT_VALUE = 2;
+
+FFilterPredicate NoFilter = [](const Worker_EntityId&, const EntityViewElement&) {
+	return true;
+};
+FComponentChangeRefreshPredicate NoComponentChangeRefreshPredicate = [](const FEntityComponentChange&) {
+	return true;
+};
+TArray<FDispatcherRefreshCallback> NoRefreshCallbacks = TArray<FDispatcherRefreshCallback>{};
+
+SUBVIEW_TEST(GIVEN_Set_Of_Components_WHEN_Entity_Tagged_THEN_Components_Contains_Tag)
+{
+	FDispatcher Dispatcher;
+	EntityView View;
+	const FSubView SubView(TAG_COMPONENT_ID, NoFilter, &View, Dispatcher, NoRefreshCallbacks);
+	TArray<FWorkerComponentData> Components{ ComponentFactory::CreateEmptyComponentData(VALUE_COMPONENT_ID) };
+
+	SubView.TagEntity(Components);
+
+	TestTrue("Components contains tag component", Components.ContainsByPredicate([](const FWorkerComponentData& Data) {
+		return Data.component_id == TAG_COMPONENT_ID;
+	}));
+
+	return true;
+}
+
+SUBVIEW_TEST(GIVEN_Query_WHEN_Query_Tagged_THEN_Query_Is_Tagged)
+{
+	FDispatcher Dispatcher;
+	EntityView View;
+	FSubView SubView(TAG_COMPONENT_ID, NoFilter, &View, Dispatcher, NoRefreshCallbacks);
+	Query QueryToTag;
+	QueryConstraint Constraint;
+	Constraint.EntityIdConstraint = ENTITY_ID;
+	QueryToTag.Constraint = Constraint;
+	QueryToTag.ResultComponentIds = SchemaResultType{ VALUE_COMPONENT_ID };
+
+	SubView.TagQuery(QueryToTag);
+
+	if (!TestNotEqual("Constraint is now an AND", QueryToTag.Constraint.AndConstraint.Num(), 0))
+	{
+		return true;
+	}
+	TArray<QueryConstraint>& AndConstraint = QueryToTag.Constraint.AndConstraint;
+	Worker_ComponentId TestConstraintId = AndConstraint[0].ComponentConstraint.IsSet() ? AndConstraint[0].ComponentConstraint.GetValue()
```

Comment:
Won't this always be in [0] ?

Reply:
I think it's [1]? Regardless, this test is asserting that the constraint is tagged correctly, not that either of the specific constraints are the component constraint. It would be valid for the function to choose to put the conjunct anywhere in the and constraint, as that would be logically the same, so that's what this tests.